kip-0015.md

```diff
@@ -0,0 +1,304 @@
+
+# Specification
+
+## Schema Sketch
+
+### Request
+```
+{
+  "commandSigDatas": [ { "cmd": string, "signatureList": [ { "publicKey":string, "signature": string|null }]}]
+}
+```
+
+### Response
+
+```
+{
+  "responses": [{
+    "commandSigData": { "cmd": string, "signatureList": [ { "publicKey":string, "signature": string|null }]},
+    "outcome": { "result": "success|failure|noSig", "hash|message"|null: string } }
+  }]
+}
+```
+or
+```
+{
+  "error": {  "type": "reject|emptyList|other", "msg": string|null }
+}
+```
+
+## **Quicksign Request**
+
+A quicksign request takes the form of a json-encoded payload matching the following schema:
+
+```
+{
+  "commandSigDatas": [CommandSigData]
+}
+```
+
+## CommandSigData
+
+- The "cmd" string represents a json-encoded payload type, specified here: [https://api.chainweb.com/openapi/pact.html#tag/model-payload](https://api.chainweb.com/openapi/pact.html#tag/model-payload)
+    - Here is an example `cmd` payload:
+
+```
+    "{\"networkId\":\"testnet04\",\"payload\":{\"exec\":{\"data\":null,\"code\":\"(+ 1 2)\"}},\"signers\":[{\"pubKey\":\"ae18efd16cbd49e9a92552a2589ac01491b486fbcbb1e3f07980f945597e2033\",\"clist\":[{\"args\":[],\"name\":\"coin.GAS\"}]},{\"pubKey\":\"fa781bdd858cd2380b5e2b654e58035f7189a6e8158686a1bb7eabb585a56e7f\"}],\"meta\":{\"creationTime\":1663085494,\"ttl\":1800,\"gasLimit\":20,\"chainId\":\"0\",\"gasPrice\":1.0e-6,\"sender\":\"ae18efd16cbd49e9a92552a2589ac01491b486fbcbb1e3f07980f945597e2033\"},\"nonce\":\"2022-09-13 16:11:34.678231 UTC\"}"
+```
+
+The schema for a `CommandSigData` object is:
+
+```
+{
+  "signatureList": [Signer],
+  "cmd": string
+}
+```
+
+Note that this type is an evolved version of `SigData` which pact currently uses in its command-line tool: [https://github.com/kadena-io/pact/blob/957b8bd7644cebc60d043b33f0d7ffc53c65d783/src/Pact/Types/SigData.hs#L62-L66](https://github.com/kadena-io/pact/blob/957b8bd7644cebc60d043b33f0d7ffc53c65d783/src/Pact/Types/SigData.hs#L62-L66)
+
+## Signer
+
+- The `publicKey` field is a lowercase base-16 encoded public key
+- The `signature` element is either a lowercase base-16 encoded signature or a null value (representing a request for signature).
+
+```
+{
+  "publicKey": string,
+  "signature": string //optional; its omission represents a signature that has not yet been acquired
+}
+```
+
+## **Quicksign Response**
+```
+{
+  "responses": [Response]
+}
+```
+
+or
+
+```
+{
+  "error": QsError
```

Comment:
I would like to suggest swapping QsError to QuicksignError.
I like keeping APIs very declarative in their types. Might be semantics, but when i first read QsError I had to think about it for a second before I knew exactly what it was referring to, which seems stupid now, but not at the time. =)

Reply:
Ah interesting. Note that `QsError` is not actually a field/key you will see in the api when using it. It is just a non-terminal symbol used to specify this (psuedo)-grammar. That being said, I wrote it to mirror the haskell implementation more, and perhaps it would make sense to switch some of the symbol's name to better the document's readability. Thanks for the suggestion!

Reply:
I disagree. 

If I'm using TypeScript or any typed language I will create a QuicksignError type that can load the JSON of the response.

And even if it is in an untyped language like Javascript or Python, you will still want to clearly define the types you might receive and give them name so developers can know precisely what to expect. 

Thus, every time you interact with Quicksign and you get an error, you must know what a QuicksignError is. 

Be rigorous and consistent in your definitions and people will like building with your stuff that much more.

Reply:
Changed